test/hashMapTest.mo

```diff
@@ -1,146 +1,518 @@
+let largeEntries = Array.tabulate<(Text, Text)>(largeSize, func i = (largeKeys[i], largeValues[i]));
+// mix in some forced collisions
+let collideHash : Text -> Hash.Hash = func key {
+  if (key < "key10") {
+    0
+  } else if (key < "key20") {
+    1
+  } else {
+    Text.hash(key)
+  }
+};
 
-  // ensure clone has each value present in original
-  for (v in a.vals()) {
-    var foundMatch = false;
-    for (w in b.vals()) {
-      if (v == w) { foundMatch := true }
-    };
-    assert foundMatch
+func newCollidedMap() : HashMap.HashMap<Text, Text> {
+  let collidedMap = HashMap.HashMap<Text, Text>(3, Text.equal, collideHash);
+  var i = 0;
+  while (i < largeSize) {
+    collidedMap.put(largeKeys[i], largeValues[i]);
+    i += 1
   };
+  collidedMap
+};
 
-  // do some more operations:
-  a.put("apple", 1111);
-  a.put("banana", 2222);
-  switch (a.remove("pear")) {
-    case null { assert false };
-    case (?three) { assert three == 3 }
-  };
-  a.delete("avocado");
+func newEmptyMap() : HashMap.HashMap<Text, Text> {
+  HashMap.HashMap<Text, Text>(1, Text.equal, Text.hash)
+};
 
-  // check them:
-  switch (a.get("apple")) {
-    case (?1111) {};
-    case _ { assert false }
-  };
-  switch (a.get("banana")) {
-    case (?2222) {};
-    case _ { assert false }
-  };
-  switch (a.get("pear")) {
-    case null {};
-    case (?_) { assert false }
-  };
-  switch (a.get("avocado")) {
-    case null {};
-    case (?_) { assert false }
-  };
-
-  // undo operations above:
-  a.put("apple", 1);
-  // .. and test that replace works
-  switch (a.replace("apple", 666)) {
-    case null { assert false };
-    case (?one) {
-      assert one == 1; // ...and revert
-      a.put("apple", 1)
+//FIXME move into Nat module in next PR
+func fromText(text : Text) : ?Nat {
+  var n = 0;
+  for (c in text.chars()) {
+    if (Char.isDigit(c)) {
+      let charAsNat = Prim.nat32ToNat(Prim.charToNat32(c) -% Prim.charToNat32('0'));
+      n := n * 10 + charAsNat
+    } else {
+      return null
     }
   };
-  a.put("banana", 2);
-  a.put("pear", 3);
-  a.put("avocado", 4);
+  ?n
+};
 
-  // ensure clone has each key-value pair present in original
-  for ((k, v) in a.entries()) {
-    Prim.debugPrint(debug_show (k, v));
-    switch (b.get(k)) {
-      case null { assert false };
-      case (?w) { assert v == w }
+// Need to sort the entries because the order is not guaranteed for equality tests
+func sortedEntries<V>(map : HashMap.HashMap<Text, V>) : [(Text, V)] {
+  Array.sort<(Text, V)>(
+    Iter.toArray(map.entries()),
+    func(p1, p2) {
+      ignore do ? {
+        let key1Index = Text.stripStart(p1.0, #text "key");
+        let key2Index = Text.stripStart(p2.0, #text "key");
+        return Nat.compare(fromText(key1Index!)!, fromText(key2Index!)!)
+      };
+      Debug.print("key1Index: " # p1.0);
+      Debug.print("key2Index: " # p2.0);
+      Debug.trap "unreachable in sortedEntries"
     }
```

Comment:
Why not just use Text.compare on the first components?

Reply:
`"10" < "2"` which was making the tests awkward.

Definitely should use Nats